mono/metadata/threads.c

```diff
@@ -4537,16 +4539,14 @@ mono_thread_execute_interruption (void)
 		UNLOCK_THREAD (thread);
 		return exc;
 	} else if (thread->state & (ThreadState_AbortRequested)) {
-		UNLOCK_THREAD (thread);
-		g_assert (sys_thread->pending_exception == NULL);
-		if (thread->abort_exc == NULL) {
-			/* 
-			 * This might be racy, but it has to be called outside the lock
-			 * since it calls managed code.
-			 */
-			MONO_OBJECT_SETREF (thread, abort_exc, mono_get_exception_thread_abort ());
+		MonoException *exc = thread->abort_exc;
```

Comment:
This will reuse the exception object which is not a good idea in most cases, since the app will except these objects to be unique.

Reply:
You can allocate the exception outside the lock, and install it inside the lock.